File: 2016/12.cs
namespace _2016;

public abstract class _12
{
    private static readonly string[] Data;

    static _12() => Data = Task.Run(() => Utils.GetInputData(12)).Result
        .Split('\n', StringSplitOptions.RemoveEmptyEntries);

    private static int ExecuteInstructions(Dictionary<string, int> registers)
    {
        var instructions = Data;
        var pointer = 0;

        while (pointer < instructions.Length)
        {
            var parts = instructions[pointer].Split(' ');

            switch (parts[0])
            {
                case "cpy":
                    if (int.TryParse(parts[1], out var value)) registers[parts[2]] = value;
                    else registers[parts[2]] = registers[parts[1]];
                    pointer++;
                    break;
                case "inc":
                    registers[parts[1]]++;
                    pointer++;
                    break;
                case "dec":
                    registers[parts[1]]--;
                    pointer++;
                    break;
                case "jnz":
                    if (int.TryParse(parts[1], out var cmpValue))
                        if (cmpValue != 0) pointer += int.Parse(parts[2]);
                        else pointer++;
                    else if (registers[parts[1]] != 0) pointer += int.Parse(parts[2]);
                        else pointer++;
                    break;
                default:
                    throw new InvalidOperationException($"Unknown instruction {parts[0]}");
            }
        }

        return registers["a"];
    }

    private static int SolvePartOne() =>
        ExecuteInstructions(new Dictionary<string, int> { { "a", 0 }, { "b", 0 }, { "c", 0 }, { "d", 0 } });

    private static int SolvePartTwo() =>
        ExecuteInstructions(new Dictionary<string, int> { { "a", 0 }, { "b", 0 }, { "c", 1 }, { "d", 0 } });

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }
}

File: 2017/12.fs
namespace _2017

open System
open System.Collections.Generic

module _12 =
    let Data = (Utils.GetInputData 12).Split('\n', StringSplitOptions.RemoveEmptyEntries)

    let buildGraph(data: string[]) =
        let graph = Dictionary<int, List<int>>()
        for line in data do
            let parts = line.Split(" <-> ", StringSplitOptions.RemoveEmptyEntries)
            let fromNode = int parts[0]
            let toNodes = parts[1].Split(',') |> Array.map int
            if not (graph.ContainsKey(fromNode)) then graph[fromNode] <- List<int>()
            for toNode in toNodes do
                graph[fromNode].Add(toNode)
                if not (graph.ContainsKey(toNode)) then graph[toNode] <- List<int>()
                graph[toNode].Add(fromNode)
        graph

    let dfs (graph: Dictionary<int, List<int>>) (visited: HashSet<int>) startNode =
        let rec visit(node: int) =
            if not (visited.Contains(node)) then
                visited.Add(node) |> ignore
                for neighbor in graph[node] do visit neighbor
        visit startNode

    let findGroupSize graph startNode =
        let visited = HashSet<int>()
        dfs graph visited startNode
        visited.Count

    let countGroups(graph: Dictionary<int, List<int>>) =
        let visited = HashSet<int>()
        graph.Keys |> Seq.fold(fun groupCount node ->
            if not (visited.Contains(node)) then
                dfs graph visited node
                groupCount + 1
            else groupCount
        ) 0

    let solvePartOne graph = findGroupSize graph 0

    let solvePartTwo graph = countGroups graph

    let Run() =
        let graph = buildGraph Data
        printfn $"Part One: {solvePartOne graph}"
        printfn $"Part Two: {solvePartTwo graph}"